### PR TITLE
Add puppet agent

### PR DESCRIPTION
**Pull request by sebastus**
@singhkay

In the absence of a Linux VM Extension for Puppet, I created this ARM template + Custom Script.  Installs the agent code on a linux box and registers the box with a puppet enterprise box.

### CODE REVIEW

**Comment by singhkay** [comment]
@sedouard Looks like the CI ran a bunch of other templates here as well. Could it because of the ones that are in the master merge?

File: linux-puppet-agent/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        "North Europe",
+        "South Central US",
+        "Southeast Asia",
+        "West Europe",
+        "West US"
+      ],
+      "metadata": {
+        "description": "The location where the Virtual Machine will be deployed, e.g: West US."
+      },
+      "defaultValue": "West US"
```
Comment (singhkay) [inline linux-puppet-agent/azuredeploy.json]: Please remove the default location

File: linux-puppet-agent/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+      "metadata": {
+        "description": "Username for the Virtual Machine, e.g: joetheadmin."
+      }
+    },
+    "password": {
+      "type": "securestring",
+      "metadata": {
+        "description": "Password for the Virtual Machine."
+      },
+      "defaultValue": "GEN-PASSWORD"
```
Comment (singhkay) [inline linux-puppet-agent/azuredeploy.json]: Let's just keep this in the parameters file for CI purposes

File: linux-puppet-agent/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    },
+    "newStorageAccountName": {
+      "type": "string",
+      "metadata": {
+        "description": "Unique DNS prefix for the Storage Account where the Virtual Machine's disks will be placed."
+      },
+      "defaultValue": "GEN-UNIQUE"
```
Comment (singhkay) [inline linux-puppet-agent/azuredeploy.json]: GEN-UNIQUE is only needed for parameters file

File: linux-puppet-agent/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+        "description": "Unique DNS Name for the Public IP used to access the Virtual Machine."
+      },
+      "defaultValue": "GEN-UNIQUE-13"
+    },
+    "fileUris": {
+      "type":"string",
+      "metadata": {
+        "description": "URI of the shell script, e.g: https://myblobs.blob.core.windows.net/public/install_puppet_agent.sh"
+      },
+      "defaultValue": "https://raw.githubusercontent.com/Azure/ARM/master/linux-puppet-agent/install_puppet_agent.sh"
```
Comment (singhkay) [inline linux-puppet-agent/azuredeploy.json]: Can you host this script in this repo and make this a variable

File: linux-puppet-agent/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+        "description": "Unique DNS prefix for the Storage Account where the Virtual Machine's disks will be placed."
+      },
+      "defaultValue": "GEN-UNIQUE"
+    },
+    "dnsNameForPublicIP": {
+      "type": "string",
+      "metadata": {
+        "description": "Unique DNS Name for the Public IP used to access the Virtual Machine."
+      },
+      "defaultValue": "GEN-UNIQUE-13"
```
Comment (singhkay) [inline linux-puppet-agent/azuredeploy.json]: This isn't needed here either

**Comment by sebastus** [comment]
@singhkay - back to you.  thanks for the comments.  the contribution guide wasn't 100% clear.

**Comment by sedouard** [comment]
@sebastus @singhkay  #959 keeps biting us in the butt HARD. Unfortunately it rears its ugly head on PRs that are based off of older versions of the repo.

Sebastus, since you did the right thing and created a branch before sending your pr you can easily rebase this PR and the CI system will only pickup your changes. If you can just go into your git console and do the following. I'm assuming you made no commits directly to your local master:

```
git checkout master
git remote add upstream https://github.com/Azure/azure-quickstart-templates.git
git fetch upstream master
git checkout AddPuppetAgent
# hopefully you have no conflicts, if you do, resolve them
git rebase master
# now push to this branch and the PR will update
git push origin AddPuppetAgent -f
```

**Comment by singhkay** [comment]
@sedouard My bad, shouldn't have merged that one :)

File: linux-puppet-agent/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+    "subnetPrefix": "10.0.0.0/24",
+    "storageAccountType": "Standard_LRS",
+    "publicIPAddressName": "[concat(variables('scenarioPrefix'),'PublicIp')]",
+    "publicIPAddressType": "Dynamic",
+    "vmStorageAccountContainerName": "vhds",
+    "vmName": "[concat(variables('scenarioPrefix'),'VM')]",
+    "virtualNetworkName": "[concat(variables('scenarioPrefix'),'Vnet')]",
+    "vnetID": "[resourceId('Microsoft.Network/virtualNetworks',variables('virtualNetworkName'))]",
+    "subnetRef": "[concat(variables('vnetID'),'/subnets/',variables('subnetName'))]",
+    "fileUris":"https://raw.githubusercontent.com/Azure/ARM/master/linux-puppet-agent/install_puppet_agent.sh",
```
Comment (sedouard) [inline linux-puppet-agent/azuredeploy.json]: This link is incorrect. Should be
`https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/linux-puppet-agent/install_puppet_agent.sh`

**Comment by sedouard** [comment]
@sebastus here's the error I managed to pull it from the portal before it got deleted:

```
{"status":"Failed","error":{"code":"ResourceDeploymentFailure","message":"The resource operation completed with terminal provisioning state 'Failed'.","details":[{"code":"VMExtensionProvisioningError","message":"VM has reported a failure when processing extension 'installcustomscript'. Error message: \"Enable failed: HTTP Error 404: Not Found\"."}]}}
``` 

Probably means your extension script doesn't have the right link

File: linux-puppet-agent/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+    "subnetPrefix": "10.0.0.0/24",
+    "storageAccountType": "Standard_LRS",
+    "publicIPAddressName": "[concat(variables('scenarioPrefix'),'PublicIp')]",
+    "publicIPAddressType": "Dynamic",
+    "vmStorageAccountContainerName": "vhds",
+    "vmName": "[concat(variables('scenarioPrefix'),'VM')]",
+    "virtualNetworkName": "[concat(variables('scenarioPrefix'),'Vnet')]",
+    "vnetID": "[resourceId('Microsoft.Network/virtualNetworks',variables('virtualNetworkName'))]",
+    "subnetRef": "[concat(variables('vnetID'),'/subnets/',variables('subnetName'))]",
+    "fileUris":"https://raw.githubusercontent.com/Azure/master/linux-puppet-agent/install_puppet_agent.sh",
```
Comment (sedouard) [inline linux-puppet-agent/azuredeploy.json]: Should be: `https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/linux-puppet-agent/install_puppet_agent.sh`

**Comment by sedouard** [comment]
Hey you passed! 🍻🍻🍻

**Comment by sebastus** [comment]
Woot!

**Comment by singhkay** [comment]
mzu0v7-i

Merging

**Pull request merged**